src/controllers/dynakube/version/updater_test.go

```diff
@@ -0,0 +1,280 @@
+		assert.Equal(t, dynatracev1beta1.CustomImageVersionSource, target.Source)
+		assertVersionStatusEquals(t, registry, testImage, *target)
+	})
+	t.Run("DON'T set source and probe at the end, if error", func(t *testing.T) {
+		registry := newEmptyFakeRegistry()
+		target := &dynatracev1beta1.VersionStatus{}
+		versionReconciler := Reconciler{
+			dynakube:     &dynatracev1beta1.DynaKube{},
+			timeProvider: timeProvider,
+			hashFunc:     registry.ImageVersionExt,
+		}
+		updater := newCustomImageUpdater(target, testImage.Uri())
+		err := versionReconciler.run(updater, testDockerCfg)
+		require.Error(t, err)
+		assert.Nil(t, target.LastProbe)
+		assert.Empty(t, target.Source)
+	})
+	t.Run("autoUpdate disabled, runs if status is empty or source changes", func(t *testing.T) {
+		registry := newEmptyFakeRegistry()
+		target := &dynatracev1beta1.VersionStatus{}
+		versionReconciler := Reconciler{
+			dynakube:     &dynatracev1beta1.DynaKube{},
+			timeProvider: timeProvider,
+			hashFunc:     registry.ImageVersionExt,
+		}
+		updater := newDefaultUpdater(target, false)
+
+		// 1. call => status empty => should run
+		err := versionReconciler.run(updater, testDockerCfg)
+		require.NoError(t, err)
+		updater.AssertNumberOfCalls(t, "UseDefault", 1)
+		assert.Equal(t, timeProvider.Now(), target.LastProbe)
+		assert.Equal(t, dynatracev1beta1.DefaultVersionSource, target.Source)
+
+		// 2. call => status NOT empty => should NOT run
+		err = versionReconciler.run(updater, testDockerCfg)
+		require.NoError(t, err)
+		updater.AssertNumberOfCalls(t, "UseDefault", 1)
+
+		// 3. call => source is different => should run
+		target.Source = dynatracev1beta1.CustomImageVersionSource
+		err = versionReconciler.run(updater, testDockerCfg)
+		require.NoError(t, err)
+		updater.AssertNumberOfCalls(t, "UseDefault", 2)
+
+		// 4. call => source is NOT different => should NOT run
+		err = versionReconciler.run(updater, testDockerCfg)
+		require.NoError(t, err)
+		updater.AssertNumberOfCalls(t, "UseDefault", 2)
+	})
+	t.Run("public registry, version set to imageTag", func(t *testing.T) {
+		registry := newFakeRegistryForImages(testImage)
+		target := &dynatracev1beta1.VersionStatus{
+			Source: dynatracev1beta1.DefaultVersionSource,
+		}
+		versionReconciler := Reconciler{
+			dynakube:     enablePublicRegistry(&dynatracev1beta1.DynaKube{}),
+			timeProvider: timeProvider,
+			hashFunc:     registry.ImageVersionExt,
+		}
+		updater := newPublicRegistryUpdater(target, &testImage, false)
+
+		err := versionReconciler.run(updater, testDockerCfg)
+		require.NoError(t, err)
+		updater.AssertNumberOfCalls(t, "LatestImageInfo", 1)
+		assert.Equal(t, timeProvider.Now(), target.LastProbe)
+		assert.Equal(t, dynatracev1beta1.PublicRegistryVersionSource, target.Source)
+		assertVersionStatusEquals(t, registry, testImage, *target)
+		assert.Equal(t, target.ImageTag, target.Version)
+	})
+}
+
+func TestDetermineSource(t *testing.T) {
+	customImage := "my.special.image"
```

Comment:
same as for the other comment about table tests, but here there is no more stuff to abstract as there is 3 simple lines of setup which differs per test case, 1 execution, 1 assert

and from previous experience the fact I would have to define a specific struct + big map, the space/code I would safe is 0 or I would even produce more lines

Reply:
It's not only about the mere number of code lines, it's also about readability and ease of grasping.

Reply:
it is a matter of taste, to me table tests are kinda unreadable because:

for example: in case of a table test to fully grasp what is happening you have to
- look at line 10 where the test case is defined
- look at line 200 where the common logic is 
so to look through a couple of test cases, you either have to memorise the common logic, so reading the test case definitions make sense, or scroll a lot 😅 


while in the current case I look at in one place, to have an understanding of what is happening